nflog/nflog.go

```diff
@@ -0,0 +1,566 @@
+package nflog
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/matttproud/golang_protobuf_extensions/pbutil"
+	pb "github.com/prometheus/alertmanager/nflog/nflogpb"
+	"github.com/weaveworks/mesh"
+)
+
+var ErrNotFound = errors.New("not found")
+
+// Log stores and serves information about notifications
+// about byte-slice addressed alert objects to different receivers.
+type Log interface {
+	// The Log* methods store a notification log entry for
+	// a fully qualified receiver and a given IDs identifying the
+	// alert object.
+	LogActive(r *pb.Receiver, key, hash []byte) error
+	LogResolved(r *pb.Receiver, key, hash []byte) error
+
+	// Query the log along the given Paramteres.
+	//
+	// TODO(fabxc):
+	// - extend the interface by a `QueryOne` method?
+	// - return an iterator rather than a materialized list?
+	Query(p ...QueryParam) ([]*pb.Entry, error)
+
+	// Snapshot the current log state and return the number
+	// of bytes written.
+	Snapshot(w io.Writer) (int, error)
+	// GC removes expired entries from the log. It returns
+	// the total number of deleted entries.
+	GC() (int, error)
+}
+
+// query currently allows filtering by and/or receiver group key.
+// It is configured via QueryParameter functions.
+//
+// TODO(fabxc): Future versions could allow querying a certain receiver
+// group or a given time interval.
+type query struct {
+	recv     *pb.Receiver
+	groupKey []byte
+}
+
+// QueryParam is a function that modifies a query to incorporate
+// a set of parameters. Returns an error for invalid or conflicting
+// parameters.
+type QueryParam func(*query) error
```

Comment:
It feels weird to me that this is an exported type with an unexported type in its definition. While it probably works as intended, `QueryParam` doesn't show up in the godoc, so the doc comment above is not visible.
How about having a struct that has the function as an unexported field? Then it would show up in godoc with the famous "// contains filtered or unexported fields".